safehttp/plugins/csp/csp_test.go

```diff
@@ -0,0 +1,236 @@
+			wantString: "frame-ancestors 'self'; report-uri httsp://example.com/collector",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := tt.policy.serialize("super-secret")
+
+			if s != tt.wantString {
+				t.Errorf("tt.policy.serialize() got: %q want: %q", s, tt.wantString)
+			}
+		})
+	}
+}
+
+func TestBefore(t *testing.T) {
+	tests := []struct {
+		name                  string
+		interceptor           Interceptor
+		wantEnforcementPolicy []string
+		wantReportOnlyPolicy  []string
+		wantNonce             string
+	}{
+		{
+			name:        "No policies",
+			interceptor: Interceptor{},
+			wantNonce:   "KSkpKSkpKSkpKSkpKSkpKSkpKSk=",
+		},
+		{
+			name:        "Default policies",
+			interceptor: Default(""),
+			wantEnforcementPolicy: []string{
+				"object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-KSkpKSkpKSkpKSkpKSkpKSkpKSk='; base-uri 'none'",
+				"frame-ancestors 'self'",
+			},
+			wantNonce: "KSkpKSkpKSkpKSkpKSkpKSkpKSk=",
+		},
+		{
+			name:        "Default policies with reporting URI",
+			interceptor: Default("https://example.com/collector"),
+			wantEnforcementPolicy: []string{
+				"object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-KSkpKSkpKSkpKSkpKSkpKSkpKSk='; base-uri 'none'; report-uri https://example.com/collector",
+				"frame-ancestors 'self'; report-uri https://example.com/collector",
+			},
+			wantNonce: "KSkpKSkpKSkpKSkpKSkpKSkpKSk=",
+		},
+		{
+			name:        "StrictCSP reportonly",
+			interceptor: NewInterceptor(StrictCSPBuilder{ReportOnly: true, ReportURI: "https://example.com/collector"}.Build()),
+			wantReportOnlyPolicy: []string{
+				"object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-KSkpKSkpKSkpKSkpKSkpKSkpKSk='; base-uri 'none'; report-uri https://example.com/collector",
+			},
+			wantNonce: "KSkpKSkpKSkpKSkpKSkpKSkpKSk=",
+		},
+		{
+			name:                 "FramingCSP reportonly",
+			interceptor:          NewInterceptor(FramingPolicy(true, "https://example.com/collector")),
+			wantReportOnlyPolicy: []string{"frame-ancestors 'self'; report-uri https://example.com/collector"},
+			wantNonce:            "KSkpKSkpKSkpKSkpKSkpKSkpKSk=",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rr := safehttptest.NewResponseRecorder()
+			req := safehttptest.NewRequest(safehttp.MethodGet, "/", nil)
+
+			tt.interceptor.Before(rr.ResponseWriter, req)
+
+			h := rr.Header()
+			if diff := cmp.Diff(tt.wantEnforcementPolicy, h.Values("Content-Security-Policy"), cmpopts.EquateEmpty()); diff != "" {
+				t.Errorf("h.Values(\"Content-Security-Policy\") mismatch (-want +got):\n%s", diff)
+			}
+
+			if diff := cmp.Diff(tt.wantReportOnlyPolicy, h.Values("Content-Security-Policy-Report-Only"), cmpopts.EquateEmpty()); diff != "" {
+				t.Errorf("h.Values(\"Content-Security-Policy-Report-Only\") mismatch (-want +got):\n%s", diff)
+			}
+
+			ctx := req.Context()
+			v := ctx.Value(ctxKey{})
```

Comment:
Are we OK with the CSP interceptor to ever *not* set a nonce?

i.e. I don't see a reason why someone couldn't write code like: `x := ctx.Value(ctxKey{}).(string)`, which would panic because the returned value would have been `nil`.

Reply:
No, not anymore. As of f2a7d01 the nonce is always added to the context. Even if the CSP doesn't use a nonce. I just forgot to propagate the change to this test. Thanks for noticing it!